pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_reconcile.go

```diff
@@ -0,0 +1,174 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiexportendpointslice
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+
+	virtualworkspacesoptions "github.com/kcp-dev/kcp/cmd/virtual-workspaces/options"
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	apiexportbuilder "github.com/kcp-dev/kcp/pkg/virtual/apiexport/builder"
+)
+
+func (c *controller) reconcile(ctx context.Context, apiExportEndpointSlice *apisv1alpha1.APIExportEndpointSlice) error {
+
+	// TODO (fgiloux): When the information is available in the cache server
+	// check if at least one APIBinding is bound in the shard to the APIExport referenced by the APIExportEndpointSLice.
+	// If so, add the respective endpoint to the status.
+	// For now the unfiltered list is added.
+
+	// Check for valid reference
+	workspaceRef := apiExportEndpointSlice.Spec.APIExport.Workspace
+	if workspaceRef == nil {
+		conditions.MarkFalse(
+			apiExportEndpointSlice,
+			apisv1alpha1.APIExportValid,
+			apisv1alpha1.APIExportInvalidReferenceReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"Missing APIExport reference",
+		)
+		return nil
+	}
+
+	// Get APIExport
+	apiExportClusterName, err := getAPIExportClusterName(apiExportEndpointSlice)
+	if err != nil {
+		conditions.MarkFalse(
+			apiExportEndpointSlice,
+			apisv1alpha1.APIExportValid,
+			apisv1alpha1.APIExportInvalidReferenceReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
+		)
+		return nil
+	}
+
+	apiExport, err := c.getAPIExport(apiExportClusterName, workspaceRef.ExportName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Don't keep the endpoints if the APIExport has been deleted
+			apiExportEndpointSlice.Status.APIExportEndpoints = nil
+			conditions.MarkFalse(
+				apiExportEndpointSlice,
+				apisv1alpha1.APIExportValid,
+				apisv1alpha1.APIExportNotFoundReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"APIExport %s|%s not found",
+				apiExportClusterName,
+				workspaceRef.ExportName,
+			)
+			return nil
+		} else {
+			conditions.MarkFalse(
+				apiExportEndpointSlice,
+				apisv1alpha1.APIExportValid,
+				apisv1alpha1.InternalErrorReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"Error getting APIExport %s|%s: %v",
```

Comment:
This error is something _other_ than `NotFound` from a lister - unclear what it might even be, but does it make sense to update status here? Can we retry the reconciliation? What might a user do with the information here?

Reply:
It's either impossible or essentially impossible to get other kinds of errors, given the current implementation of a lister. But in any case, I do think updating a condition is appropriate - it provides feedback to the user that something went wrong. They may not have any recourse, but it's better than giving them no information and having the key stuck in an endless exponential backoff loop.